test/govm-start-ubuntu.sh

```diff
@@ -0,0 +1,148 @@
+# let's have tracing via x on above line, while we debug/develop this
+
+# Catch errors in pipelines
+set -o pipefail
+
+. $(dirname $0)/govm-start-common.sh
+#_ubunturelease=bionic
+_ubunturelease=cosmic
+_imgfile=${_ubunturelease}-server-cloudimg-amd64.img
+_mirror=
+
+set_apt_mirror () {
+    # override all repos. This is called only if mirror value exists
+    a=$1
+    if [ -f _work/govm_set_apt_mirror.$a.done ]; then
+        return 0
+    fi
+    $_ssh root@$a "cat > /etc/apt/sources.list" <<EOF
+deb $_mirror ${_ubunturelease} main restricted
+deb $_mirror ${_ubunturelease}-updates main restricted
+deb $_mirror ${_ubunturelease} universe
+deb $_mirror ${_ubunturelease}-updates universe
+deb $_mirror ${_ubunturelease} multiverse
+deb $_mirror ${_ubunturelease}-updates multiverse
+deb $_mirror ${_ubunturelease}-backports main restricted universe multiverse
+deb $_mirror ${_ubunturelease}-security main restricted
+deb $_mirror ${_ubunturelease}-security universe
+deb $_mirror ${_ubunturelease}-security multiverse
+EOF
+    touch _work/govm_set_apt_mirror.$a.done
+}
+
+set_apt_proxy () {
+    # proxy for apt
+    a=$1
+    if [ -f _work/govm_apt_proxy.$a.done ]; then
+        return 0
+    fi
+    $_ssh root@$a "cat >/etc/apt/apt.conf.d/00proxy" <<EOF
+Acquire::http::Proxy "$http_proxy";
+EOF
+    # ... with exception for mirror
+    if [ -n "$_mirror" ]; then
+        $_ssh root@$a "cat >>/etc/apt/apt.conf.d/00proxy" <<EOF
+Acquire::http::Proxy::linux-ftp.fi.intel.com  "DIRECT";
+EOF
+    fi
+    touch _work/govm_apt_proxy.$a.done
+}
+
+install_kubeadm () {
+    # add Kubernetes key and repo. bionic,cosmic not available so we use latest=xenial
+    a=$1
+    if [ -f _work/govm_install_kubeadm.$a.done ]; then
+	return 0
+    fi
+    $_ssh root@$a "https_proxy=$http_proxy curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add"
+    $_ssh root@$a "apt-add-repository \"deb http://apt.kubernetes.io/ kubernetes-xenial main\""
+    $_ssh root@$a "apt install kubeadm -y"
+    touch _work/govm_install_kubeadm.$a.done
+}
+
+install_docker () {
+    # note we also install ndctl here, reusing apt install point for convenience
+    a=$1
+    if [ -f _work/govm_install_docker.$a.done ]; then
+	return 0
+    fi
+    $_ssh root@$a "apt install ndctl docker.io -y"
+    touch _work/govm_install_docker.$a.done
+}
+
+final_tuning () {
+    # some misc. operations
+    a=$1
+    if [ -f _work/govm_final_tuning.$a.done ]; then
+	return 0
+    fi
+    $_ssh root@$a "systemctl enable docker.service"
+    # silence timesync, fails to connect outside. Other way would be to configure ntp server
+    $_ssh root@$a "systemctl stop systemd-timesyncd"
+    touch _work/govm_final_tuning.$a.done
+}
+
+########################  START   ###################################
+# get image, checking is it cached
+mkdir -p _work
+if [ ! -e $_imgdir/$_imgfile ]; then
+    curl -Ok https://cloud-images.ubuntu.com/${_ubunturelease}/current/$_imgfile
```

Comment:
Allow insecure connections?! Please don't, not even for development.

I prefer long options in scripts for non-obvious flags, it's easier to review. `--insecure` would have raised a red flag immediately while `-k` is easy to overlook.

Add `--location` (allow redirects) and `--fail` (fail with an error when the URL leads to 404). I know, the `curl` invocation in `test/clear-kvm.make` doesn't have it either. Your chance to deliver something better :grin:

Reply:
that cmd I copied out from govm use example on CI side. lets improve